### ISSUE TITLE
noSmooth does not work in P2D / P3D

### ISSUE THREAD

**Issue by clankill3r** [processing/processing#5363]
Processing 3.3.6
Mac OSX 10.12.6

noSmooth does not work with P2D or P3D.

<img width="752" alt="screen shot 2018-01-10 at 13 50 13" src="https://user-images.githubusercontent.com/738650/34773800-3f03cd62-f60d-11e7-922b-a291249527ae.png">

```java

PGraphics pg;

void setup() {
  size(640, 640, P2D);
  noSmooth();
  pg = createGraphics(32, 32);
  pg.noSmooth();
}

void draw() {
  pg.beginDraw();
  pg.background(128);
  pg.rect(4, 4, 16, 16);
  pg.endDraw();
  image(pg, 0, 0, width, height);
}
```

**Comment by gohai** [comment]
[This thread](https://forum.processing.org/two/discussion/24084/nosmooth-isnt-working-in-p3d) seems to indicate that `((PGraphicsOpenGL)g).textureSampling(3);` does the trick - perhaps `noSmooth` could be made to do this under the hood on the P2D renderer?

**Comment by clankill3r** [comment]
Nice. It can also be used after setup.

**Comment by benfry** [comment]
Hm, it should be working, so long as it's in `settings()` (or `setup()` in the PDE).

**Comment by JakubValtar** [comment]
`noSmooth()` in OpenGL sets only multisampling, not texture filtering. We currently don't have public API to set texture filtering.

**Comment by benfry** [comment]
But in 2D, `noSmooth()` affects images as well, so the expectation would be the same for OpenGL. (Maybe I'm misunderstanding something?)

**Comment by JakubValtar** [comment]
I think in Java2D it goes to nearest neighbor, because `noSmooth()` was used to improve performance and turning off image interpolation helped a lot.

When we got OpenGL, texture filtering was left separate, because it has negligible impact on performance, but lowers image quality a lot compared to AA. Users usually don't want to get pixellated textures, unless it is for artistic reasons (and those would probably want to leave AA on).

Other reason is that `smooth()`/`noSmooth()` can be called only once per sketch, but it should be possible to change texture filtering multiple times. Plus FX does not even support nearest neighbor.

If we change the behavior, I suggest separating these two in Java2D, because they are two different things used for different purposes (although in Java2D the performance effect overlaps).

**Comment by benfry** [comment]
Nah, it's implemented in Java2D for the effect of it, more than anything else. Things look crummy enough (and performance isn't the factor) that it's really the only reason to do `noSmooth()` nowadays. I know @codeanticode and I discussed a separate `imageSmooth()` method but at the end of the day decided to keep things as one. I suppose we could revisit the discussion…

But at the moment, we're still dealing with buggy behavior in terms of expectations. I say `noSmooth()` at the beginning of my sketch, and then I see (ugly) smoothed gradients on images, I'm gonna be disappointed (as was the OP).

**Comment by JakubValtar** [comment]
There is `noSmooth()` override (which should implement this) commented out
in `PGraphicsOpenGL`. That's probably the problem. I might check it out
tomorrow/Sunday.

**Comment by eulphean** [comment]
((PGraphicsOpenGL)g).textureSampling(3) doesn't seem to help when I draw an ellipse. Why is that so? 
PGraphics pg;

void setup() {
  size(640, 640, P2D);
  noSmooth();
  ((PGraphicsOpenGL)g).textureSampling(3);
  pg = createGraphics(32, 32);
  pg.smooth();
}

void draw() {
  pg.beginDraw();
  pg.background(128);
  pg.ellipseMode(CENTER);
  pg.ellipse(16, 16, 10, 10);
  pg.endDraw();
  image(pg, 0, 0, width, height);
}

**Comment by chrsmlls333** [comment]
Note: in P2D, this textureSampling(3) has no effect when drawing a PGraphics canvas object with image(). For example zooming in 2x on a surface half the size of our main canvas results in this same smoothing issue. Any tips?

**Comment by clankill3r** [comment]
@chrsmlls333 can you post an example?

**Comment by OpenSauce04** [comment]
This is still an issue 5 years later

**Comment by clankill3r** [comment]
@aphid91 I have limited time, but can you try to call `gl_nearest_for_texture` for your `target` PGraphics?


```
public void gl_nearest_for_texture(PImage image) {
  PGL pgl = beginPGL();
  Texture image_tex = ((PGraphicsOpenGL)g).getTexture(image);
  pgl.bindTexture(PGL.TEXTURE_2D, image_tex.glName);
  pgl.texParameteri(PGL.TEXTURE_2D, PGL.TEXTURE_MIN_FILTER, PGL.NEAREST);
  pgl.texParameteri(PGL.TEXTURE_2D, PGL.TEXTURE_MAG_FILTER, PGL.NEAREST);
  pgl.bindTexture(PGL.TEXTURE_2D, 0);
  endPGL();
}


public void gl_linear_for_texture(PImage image) {
  PGL pgl = beginPGL();
  Texture image_tex = ((PGraphicsOpenGL)g).getTexture(image);
  pgl.bindTexture(PGL.TEXTURE_2D, image_tex.glName);
  pgl.texParameteri(PGL.TEXTURE_2D, PGL.TEXTURE_MIN_FILTER, PGL.LINEAR);
  pgl.texParameteri(PGL.TEXTURE_2D, PGL.TEXTURE_MAG_FILTER, PGL.LINEAR);
  pgl.bindTexture(PGL.TEXTURE_2D, 0);
  endPGL();
}

public Texture gl_texture(PImage img) {
  return ((PGraphicsOpenGL)g).getTexture(img);
}

static public Texture gl_texture(PApplet p, PImage img) {
  return ((PGraphicsOpenGL)p.g).getTexture(img);
}
```

**Comment by aphid91** [comment]
@clankill3r I tried calling gl_nearest_for_texture(target) right after the call to target.noSmooth(), and the artifact problem persists. I also tried it with plain old noSmooth() instead of/ in addition to target.noSmooth() and the results are the same. A couple notes: 
- The code I posted only shows the test pattern if run from processing 4. Expected (though still buggy) behaivior can be restored in processing 3 by adding target.beginDraw(); target.endDraw() right before drawTestPattern(); P3 is a little weird about PGraphics initialization. I have edited my original post to reflect this.
- Similarly, in Processing 4, calling gl_nearest_for_texture(target) before the first target.beginDraw(); target.endDraw() pair causes the console to complain that "pixels array is null". Adding a beginDraw-endDraw pair right before the function call has expected behavior (still full of artifacts)
- gl_nearest_for_texture(target) must be called after target.noSmooth() as it contains a "beginDraw()"

**Comment by IntangibleMatter** [comment]
Remains an issue in 2025, making it really hard to have pixel graphics and shaders at once.
